hydra-node/test/Hydra/Chain/Direct/StateSpec.hs

```diff
@@ -144,13 +178,157 @@ spec = parallel $ do
+            let toReplay = blks \\ [blk | blk <- blks, blockPoint blk <= rollbackPoint]
+            st'' <- run $ mapM_ (onRollForward handler) toReplay *> readTVarIO headState
+            assert (st' == st'')
+
+withCounterExample :: [Block] -> TVar IO SomeOnChainHeadStateAt -> IO a -> PropertyM IO a
+withCounterExample blks headState step = do
+  stBefore <- run $ readTVarIO headState
+  a <- run step
+  stAfter <- run $ readTVarIO headState
+  a <$ do
+    monitor $
+      counterexample $
+        toString $
+          unlines
+            [ "Head state at (before rollback): " <> showStateRecordedAt stBefore
+            , "Head state at (after rollback):  " <> showStateRecordedAt stAfter
+            , "Block sequence: \n"
+                <> unlines
+                  ( fmap
+                      ("    " <>)
+                      [show (blockPoint blk) | blk <- blks]
+                  )
+            ]
+
+genRollbackPoint :: [Block] -> Gen (Word, Point Block)
+genRollbackPoint blks = do
+  let maxSlotNo = blockSlotNo (Prelude.last blks)
+  ix <- SlotNo <$> choose (1, unSlotNo maxSlotNo)
+  let rollbackDepth = length blks - length [blk | blk <- blks, blockSlotNo blk <= ix]
+  rollbackPoint <- blockPoint <$> genBlockAt ix []
+  pure (fromIntegral rollbackDepth, rollbackPoint)
+
+-- | Generate a non-sparse sequence of blocks each containing an observable
+-- transaction, starting from the returned on-chain head state.
+--
+-- Note that this does not generate the entire spectrum of observable
+-- transactions in Hydra, but only init and commits, which is already sufficient
+-- to observe at least one state transition and different levels of rollback.
+genSequenceOfObservableBlocks :: Gen (SomeOnChainHeadStateAt, [Block])
+genSequenceOfObservableBlocks = do
+  ctx <- genHydraContext 3
+
+  -- NOTE: commits must be generated from each participant POV, and thus, we
+  -- need all their respective StIdle to move on.
+  let stIdles = flip map (zip (ctxVerificationKeys ctx) (ctxParties ctx)) $ \(vk, p) ->
+        let peerVerificationKeys = ctxVerificationKeys ctx \\ [vk]
+         in idleOnChainHeadState (ctxNetworkId ctx) peerVerificationKeys vk p
+
+  stIdle <- elements stIdles
+  blks <- flip execStateT [] $ do
+    initTx <- stepInit ctx stIdle
+    void $ stepCommits ctx initTx stIdles
+
+  pure (stAtGenesis (SomeOnChainHeadState stIdle), reverse blks)
+ where
+  nextSlot :: Monad m => StateT [Block] m SlotNo
+  nextSlot = do
+    get <&> \case
+      [] -> 1
+      x : _ -> SlotNo . succ . unSlotNo . blockSlotNo $ x
+
+  putNextBlock :: Tx -> StateT [Block] Gen ()
+  putNextBlock tx = do
+    sl <- nextSlot
+    blk <- lift $ genBlockAt sl [tx]
+    modify' (blk :)
+
+  stepInit ::
+    HydraContext ->
+    OnChainHeadState StIdle ->
+    StateT [Block] Gen Tx
+  stepInit ctx stIdle = do
+    initTx <-
+      lift $
+        initialize (ctxHeadParameters ctx) (ctxVerificationKeys ctx)
+          <$> genTxIn
```

Comment:
Only `genTxIn` is monadic and needs `lift`. Maybe easier to read if we do `txIn <- lift $ genTxIn` before the pure `initialize`.